### ISSUE TITLE
Cyclical behavior of tarred ASR training data sets

### ISSUE THREAD

**Issue by shanecarroll-smarsh** [NVIDIA/NeMo#6084]
**Describe the bug**

Using tarred buckets with ASR data exhibits cyclical behavior that makes it difficult for ASR models to converge in some situations.
Not sure if I'm using it wrong, or if this is be design.

I believe the problem comes from here:

https://github.com/NVIDIA/NeMo/blob/d06d353ae3ce1224d95dc3cdeffb987f66d911ab/nemo/collections/asr/data/audio_to_text.py#L1286-L1287

For large ASR training sets, I use tarred buckets with the typical 8 buckets up to ~18 seconds max. This issue happens across many languages with a similar setup.

<details>
  <summary>Example config</summary>

  ```yaml

  train_ds:
    sample_rate: 16000
    batch_size: 1  # 1 because we're using tarred datasets
    shuffle: true
    num_workers: 8
    pin_memory: true
    use_start_end_token: false
    trim_silence: false
    max_duration: 20.0
    min_duration: 1.0
    shuffle_n: 512
    is_tarred: true
    manifest_filepath:
      - - /path/to/bucket1/tarred_audio_manifest.json
      - - /path/to/bucket2/tarred_audio_manifest.json
      - - /path/to/bucket3/tarred_audio_manifest.json
      - - /path/to/bucket4/tarred_audio_manifest.json
      - - /path/to/bucket5/tarred_audio_manifest.json
      - - /path/to/bucket6/tarred_audio_manifest.json
      - - /path/to/bucket7/tarred_audio_manifest.json
      - - /path/to/bucket8/tarred_audio_manifest.json
    tarred_audio_filepaths:
      - - /path/to/bucket1/audio_{0..511}.tar
      - - /path/to/bucket2/audio_{0..511}.tar
      - - /path/to/bucket3/audio_{0..511}.tar
      - - /path/to/bucket4/audio_{0..511}.tar
      - - /path/to/bucket5/audio_{0..511}.tar
      - - /path/to/bucket6/audio_{0..511}.tar
      - - /path/to/bucket7/audio_{0..511}.tar
      - - /path/to/bucket8/audio_{0..511}.tar
    bucketing_batch_size: [44, 26, 18, 14, 11, 9, 8, 7]
  ```
</details>

Fine-tuning a large Conformer, this is what I see when evaluating every 1/4 epoch (domain fine-tuning, expect a couple point drop in WER):

![image](https://user-images.githubusercontent.com/103518212/220742240-dd40c2c2-b254-4868-8124-15e5cedec192.png)

Since this is a fine-tuning, so a low LR is used (effective LR on the order of 1e-5), and LR is unlikely the problem w.r.t. convergence. 

Re-scaling shows that the two datasets are negatively correlated, as the `RandomizedChainDataset` tunes the model to different buckets:

![image](https://user-images.githubusercontent.com/103518212/220743868-6d321c96-31f6-4559-bb1d-5d593bf9e88d.png)

I suspect this is because `RandomizedChainDataset` exhausts each bucket in turn (as linked above) and does not re-shuffle at each epoch, and the training set duration are clustered as shown here (four common, large, publicly-available datasets):

![image](https://user-images.githubusercontent.com/103518212/220743372-dc42bea8-adb7-41db-99c6-abc87c14c3bc.png)

So different buckets are dominated by different datasets, and exhausting a bucket without training from the others fine-tunes the model to a particular dataset.

**Expected behavior**

* Sample from a random bucket each step, rather than exhausting each bucket in turn.
* Use different order each epoch.

**Comment by titu1994** [comment]
Bucketing is used for datasets in order of thousands of hours of speech. Heuristically, id put tarred datasets at < 10,00 hours, and bucketing beyond that. Also, you should use smaller bucks and number of tarfiles per buckets so that it does not exhaust data so quickly. 

Resetting the dataloaders is necessary to avoid NCCL hanging due to some nodes finishing their work vs others not finishing yet. 

As to reshuffling, we normally expect WebDataset to do that, we will look into why that's happening. 

Fyi @VahidooX @bmwshop

**Comment by shanecarroll-smarsh** [comment]
ok thanks, I'll go back to using manifest-based data sets. I use tarred buckets because the length bucketing significantly reduces training time, but if these caveats are unavoidable then it's better to train slower.

BTW this is on the order of 1000's of hours of data; the fine-tuning is synthetic in form so a lot of data can be used.

**Comment by VahidooX** [comment]
If you use synced_randomized or fully_randomized, the order of buckets at each epoch would get shuffled. You may print the the dataset_idx in the iter method of RandomizedChainDataset to check and verify it. It it is not the case, then there is a bug. 

One difference between the synced_randomized and fully_randomized is that in fully_randomized, each GPU also uses different orders. Have you tried fully_randomized?

BTW, the lr 1e-5 looks too low to me. Have you tried larger lr?

How may tarred files do you have in each bucket? How many GPUs in total?

Due to the limitations of webdataset and in order to be efficient, each GPU would consume the buckets one by one. The order of buckets are different at each epoch. In synced_randomized, at each step all GPUs uses the same bucket while with fully randomized, each GPU uses a random bucket. So if you use fully randomized with multi-GPU you should not have any issue. If you use just one GPU, it is tricky as synced and fully randomized would be the same on single GPU.

**Comment by shanecarroll-smarsh** [comment]
The nominal LR is much higher: 0.1. The effective LR consider annealing and warmup does not exceed 7e-5, but this is correct for the task.

These results are from single-GPU training, and the default strategy was used (I believe `synced_randomized`). 512 tar files per bucket, 8 buckets, 1000's of hours of audio total.

I presume this problem is less noticeable with more GPUs, as the buckets would be used in parallel. For single-GPU users, it looks like the current implementation should be avoided.
